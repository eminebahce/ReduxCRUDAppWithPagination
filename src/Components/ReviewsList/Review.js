import React from 'react'
import {
    Button,
    ListItem,
    ListItemAvatar,
    ListItemText,
    Avatar,
    Typography,
    IconButton,
    Grid,
    Tooltip
} from '@material-ui/core'
import Divider from '@material-ui/core/Divider';
import Delete from '@material-ui/icons/Delete'
import StarRatingComponent from 'react-star-rating-component';
import {deleteReview, loadReviews} from '../../Actions/ReviewActions';
import {connect} from "react-redux";
import PropTypes from 'prop-types';

class Review extends React.Component {

    onDelete = (reviewId) => {
        if (this.props.reviewSize !== 1 || this.props.totalCount === 0) {
            this.props.deleteReview(reviewId, this.calculateCurrentPage());
        } else {
            this.props.deleteReview(reviewId, this.calculatePreviousPage());
        }
    }

    calculateCurrentPage = () => {
        const firstSplit = this.props.currentUrl.split("&");
        const secondSplit = firstSplit[0].split("=");
        return parseInt(secondSplit[1]);
    }

    calculatePreviousPage = () => {
        return this.calculateCurrentPage() - 1;
    }

    handleClick = (url) => {
        this.props.loadReviews(url);
    }

    render() {
        return (
            <div data-test="review">
                {!this.props.reviews && <p>Loading...</p>}
                <div>
                    {this.props.reviews && this.props.reviews.map((review, index) =>
                        <div key={index}>
                            <ListItem alignItems="flex-start">
                                <ListItemAvatar>
                                    <Avatar alt="avatar"
                                            src="/outline-account_circle-black-18/2x/outline_account_circle_black_18dp.png"/>
                                </ListItemAvatar>
                                <ListItemText
                                    primary={
                                        <React.Fragment>
                                            <StarRatingComponent
                                                name="star-rating"
                                                value={review.rating}
                                                editing={false}
                                            />
                                        </React.Fragment>
                                    }
                                    secondary={
                                        <React.Fragment>
                                                <Typography component="span" style={{display: 'inline'}}
                                                            color="textPrimary">
                                                    {`Anonymous User`}
                                                </Typography>
                                            {` - ${review.body}`}
                                        </React.Fragment>
                                    }
                                />
                                <Tooltip title="Click to remove the review">
                                    <IconButton color='primary' onClick={() => this.onDelete(review.id)}>
                                        <Delete/>
                                    </IconButton>
                                </Tooltip>
                            </ListItem>
                            <Divider/>
                        </div>
                    )}
                </div>
                <div>
                    <Grid container
                          direction="row"
                          justify="center"
                          alignItems="center">
                        {this.props.links && Object.keys(this.props.links).map((link, index) =>
                            <Grid item key={index}
                                  role="group">
                                <Button type="button"
                                        color="primary"
                                        variant="outlined"
                                        size="small"
                                        style={{marginRight: '5px', marginTop: '10px'}}
                                        onClick={() => this.handleClick(this.props.links[link].url)}>
                                    {this.props.links[link].rel}
                                </Button>
                            </Grid>
                        )}
                    </Grid>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    reviewSize: state.reviewSize,
    totalCount: state.totalCount,
    currentUrl: state.currentUrl
})

Review.propTypes = {
    onDelete: PropTypes.func,
    calculateCurrentPage: PropTypes.func,
    calculatePreviousPage: PropTypes.func,
    handleClick: PropTypes.func
}

export default connect(mapStateToProps, {deleteReview, loadReviews})(Review);