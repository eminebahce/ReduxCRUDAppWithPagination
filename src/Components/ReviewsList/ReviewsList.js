import React, {Component} from 'react'
import {
    Grid,
    List,
    ListItem,
    Typography
} from '@material-ui/core';
import Review from './Review';
import {loadReviews} from '../../Actions/ReviewActions';
import {connect} from "react-redux";
import ReactLoading from 'react-loading';

class ReviewsList extends Component {

    state = {
        isLoading: true
    };

    componentDidMount() {
        setTimeout(() => {
            this.setState({
                isLoading: false
            })
            this.props.loadReviews();
        }, 3000);
    }

    componentWillUnmount() {
        clearTimeout();
    }

    render() {
        return (
            <div data-test="reviews-list">
                <Grid container justify="center" alignContent="center">
                    <Grid item xs={6}>
                        <List>
                            {this.state.isLoading &&
                            <ListItem>
                                <Grid container
                                      direction="column"
                                      justify="center"
                                      alignItems="center">
                                    <Grid item><ReactLoading type='spokes' color="#276190"/></Grid>
                                    <Grid item><i><Typography variant="subtitle1" color="inherit">Loading...</Typography></i></Grid>
                                </Grid>
                            </ListItem>
                            }
                            <Review reviews={this.props.reviews} links={this.props.links}/>
                        </List>
                    </Grid>
                </Grid>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    reviews: state.reviews,
    links: state.links
})
export default connect(mapStateToProps, {loadReviews})(ReviewsList);