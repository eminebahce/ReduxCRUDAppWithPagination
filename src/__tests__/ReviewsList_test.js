import React from 'react';
import {shallow} from 'enzyme';
import {Provider} from "react-redux";
import store from "../store";
import renderer from 'react-test-renderer';
import ReviewsList from '../../src/Components/ReviewsList/ReviewsList';

describe('<ReviewsList /> component', () => {
    it('should render without errors', () => {
        const component = shallow(<ReviewsList />);
        expect(component.find(`[data-test='formTextArea']`)).toBeTruthy();
    });

    describe('renders ReviewsList element', () => {
        it('should not return error', () => {
            const component = renderer.create(
                <Provider store={store}>
                    <ReviewsList />
                </Provider>
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});