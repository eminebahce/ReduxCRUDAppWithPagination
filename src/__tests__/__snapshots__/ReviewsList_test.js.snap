// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewsList /> component renders ReviewsList element should not return error 1`] = `
<div
  data-test="reviews-list"
>
  <div
    className="MuiGrid-container-1 MuiGrid-align-content-xs-center-13 MuiGrid-justify-xs-center-18"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-6-35"
    >
      <ul
        className="MuiList-root-98 MuiList-padding-99"
      >
        <li
          className="MuiListItem-root-102 MuiListItem-default-105 MuiListItem-gutters-110"
          disabled={false}
        >
          <div
            className="MuiGrid-container-1 MuiGrid-direction-xs-column-4 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18"
          >
            <div
              className="MuiGrid-item-2"
            >
              <div
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<svg id=\\"loading\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(0 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(45 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.125s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(90 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.25s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(135 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.375s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(180 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(225 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.675s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(270 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.75s\\"/>
  </path>
  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(315 16 16)\\">
    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.875s\\"/>
  </path>
</svg>
",
                  }
                }
                style={
                  Object {
                    "fill": "#276190",
                    "height": 64,
                    "width": 64,
                  }
                }
              />
            </div>
            <div
              className="MuiGrid-item-2"
            >
              <i>
                <h6
                  className="MuiTypography-root-114 MuiTypography-subtitle1-132 MuiTypography-colorInherit-143"
                >
                  Loading...
                </h6>
              </i>
            </div>
          </div>
        </li>
        <div
          data-test="review"
        >
          <div />
          <div>
            <div
              className="MuiGrid-container-1 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18"
            />
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
